hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+
+  /**
+   * Reads full or partial Chunk from DN Container based on the current
+   * position of the ChunkInputStream, the number of bytes of data to read
+   * and the checksum boundaries.
+   * If successful, then the read data in saved in the buffers so that
+   * subsequent read calls can utilize it.
+   * @param len number of bytes of data to be read
+   * @throws IOException if there is an I/O error while performing the call
+   * to Datanode
+   */
+  private synchronized void readDataFromContainer(int len) throws IOException {
+    // index of first byte to be read from the block
+    long startByteIndex;
+    if (blockPosition >= 0) {
+      // If seek operation was called to advance the buffer position, the
+      // chunk should be read from that position onwards.
+      startByteIndex = blockPosition;
+    } else {
+      // Start reading the block from the last blockPosition onwards.
+      startByteIndex = bufferOffsetWrtBlockDataData + buffersSize;
+    }
+
+    // bufferOffsetWrtChunkData and buffersSize are updated after the data
+    // is read from Container and put into the buffers, but if read fails
+    // and is retried, we need the previous position.  Position is reset after
+    // successful read in adjustBufferPosition()
+    blockPosition = getPos();
+    bufferOffsetWrtBlockDataData = readData(startByteIndex, len);
+    long tempOffset = 0L;
+    buffersSize = 0L;
+    bufferoffsets = new ArrayList<>(buffers.size());
+    for (ByteBuffer buffer : buffers) {
+      bufferoffsets.add(tempOffset);
+      tempOffset += buffer.limit();
+      buffersSize += buffer.limit();
+
+    }
+    bufferIndex = 0;
+    allocated = true;
+    adjustBufferPosition(startByteIndex - bufferOffsetWrtBlockDataData);
+
+  }
+
+  @VisibleForTesting
+  protected long readData(long startByteIndex, long len)
+      throws IOException {
+    Pipeline pipeline = pipelineRef.get();
+    buffers = new ArrayList<>();
+    ReadBlockResponseProto response =
+        ContainerProtocolCalls.readBlock(xceiverClient, startByteIndex,
+        len, blockID, validators, tokenRef.get(), pipeline.getReplicaIndexes(), verifyChecksum);
+    List<ReadChunkResponseProto> readBlocks = response.getReadChunkList();
+
+    for (ReadChunkResponseProto readBlock : readBlocks) {
+      if (readBlock.hasDataBuffers()) {
+        buffers.addAll(BufferUtils.getReadOnlyByteBuffers(
+            readBlock.getDataBuffers().getBuffersList()));
+      } else {
+        throw new IOException("Unexpected error while reading chunk data " +
+            "from container. No data returned.");
+      }
+    }
+    return response.getReadChunk(0)
+        .getChunkData().getOffset();
+  }
+
+  /**
+   * Check if the buffers have any data remaining between the current
+   * position and the limit.
+   */
+  private boolean buffersHaveData() {
+    boolean hasData = false;
+    if (buffersAllocated()) {
+      int buffersLen = buffers.size();
+      while (bufferIndex < buffersLen) {
```

Comment:
Can we change this condition to:
`while (bufferIndex < buffersLen - 1)` and remove the nested else
```
else {
            // no more buffers remaining
            break;
          }
```
 i.e the empty buffer condition else statement?

Say **buffersLen = 3** and **bufferIndex = 1.**

In `if (bufferIndex < buffersLen - 1)` this will change to bufferIndex = 2
In the next while loop we will reach the else block of the empty buffer condition and break effectively stopping the loop at the `< buffersLen - 1` condition